src/SampleLibraryUI/Examples/ButtonCustomNodeModel.cs

```diff
@@ -0,0 +1,336 @@
+
+            // This call is required to ensure that your ports are
+            // properly created.
+            RegisterAllPorts();
+
+            // Listen for input port disconnection to trigger button UI update
+            this.PortDisconnected += ButtonCustomNodeModel_PortDisconnected;
+
+            // The arugment lacing is the way in which Dynamo handles
+            // inputs of lists. If you don't want your node to
+            // support argument lacing, you can set this to LacingStrategy.Disabled.
+            ArgumentLacing = LacingStrategy.Disabled;
+
+            // We create a DelegateCommand object which will be 
+            // bound to our button in our custom UI. Clicking the button 
+            // will call the ShowMessage method.
+            ButtonCommand = new DelegateCommand(ShowMessage, CanShowMessage);
+
+            // Setting our property here will trigger a 
+            // property change notification and the UI 
+            // will be updated to reflect the new value.
+            ButtonText = defaultButtonText;
+            WindowText = defaultWindowText;
+        }
+
+        // Starting with Dynamo v2.0 you must add Json constructors for all nodeModel
+        // dervived nodes to support the move from an Xml to Json file format.  Failing to
+        // do so will result in incorrect ports being generated upon serialization/deserialization.
+        [JsonConstructor]
+        ButtonCustomNodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts)
+        {
+            this.PortDisconnected += ButtonCustomNodeModel_PortDisconnected;
+            ButtonCommand = new DelegateCommand(ShowMessage, CanShowMessage);
+        }
+
+        // Restore default button/window text and trigger UI update
+        private void ButtonCustomNodeModel_PortDisconnected(PortModel obj)
+        {
+            ButtonText = defaultButtonText;
+            WindowText = defaultWindowText;
+            RaisePropertyChanged("ButtonText");
+        }
+
+        #endregion
+
+        // Use the VMDataBridge to safely retrieve our input values
+        #region databridge callback
+        /// <summary>
+        /// Register the data bridge callback.
+        /// </summary>
+        protected override void OnBuilt()
+        {
+            base.OnBuilt();
+            VMDataBridge.DataBridge.Instance.RegisterCallback(GUID.ToString(), DataBridgeCallback);
+        }
+
+        /// <summary>
+        /// Unregister the data bridge callback.
+        /// </summary>
+        public override void Dispose()
+        {
+            base.Dispose();
+            VMDataBridge.DataBridge.Instance.UnregisterCallback(GUID.ToString());
+        }
+
+        /// <summary>
+        /// Callback method for DataBridge mechanism.
+        /// This callback only gets called once after the BuildOutputAst Function is executed 
```

Comment:
should this say, This callback only gets called when the AST is executed and only after the BuildOutputAST function is executed and the AST is fully built.... something like that?